### ISSUE TITLE
prevent inclusion of stack trace in response (leak of detailed error information)

### ISSUE THREAD

**Issue by rursprung** [opensearch-project/OpenSearch#4255]
**Is your feature request related to a problem? Please describe.**
there are currently two options to include a stack trace in the server response in case of errors:
* set the `http.detailed_errors.enabled` config option (undocumented for OpenSearch, [documented for ES 7.10.2](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/modules-http.html))
* add the `error_trace=true` URL parameter (see the [OpenSearch documentation](https://opensearch.org/docs/latest/opensearch/common-parameters/))

the former can only be done by a system administrator, however the latter can be done by anyone who can call OpenSearch.

thus, when an exception is raised, the following parameter can be added to the request URL to include the error stack trace in the response `error_trace=true`

Example Request:
```
GET /opensearch/_search?error_trace=true
```

Such technical details are not supposed to be visible to clients as it does not provide any added value. To an attacker however, this information might give valuable hints as to which frameworks and libraries are used by the web application. Apart from that, technical details can help in the analysis of errors and simplify the search for security issues.

**Describe the solution you'd like**
Because the client should only receive information that is necessary to handle the error, it should be possible to disable the corresponding parameter globally using a config option.

i could e.g. imagine changing `http.detailed_errors.enabled` from a boolean parameter to an enum in a backwards compatible way so that it accepts `false` (by default no trace is printed, the HTTP param is accepted), `true` (trace printed by default) and `disabled` (HTTP param is ignored, trace is never printed).

**Describe alternatives you've considered**
arguing with the security reviewers about the finding is usually not the best alternative 😆

**Additional context**
scoring acc. to our reviewers:

CVSS 3.1: AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:N/A:N CVSS Base Score: 5.3

note: i'm raising this as a feature request instead reporting it via email as a security issue as it is IMHO not a direct security problem and thus can be discussed openly. in a proper setup OpenSearch should also not be reachable from the outside, i.e. this can only be exploited if the attacker is already in the network.

**Comment by BhumikaSaini-Amazon** [comment]
Should this have been reported via the channels outlined in https://github.com/opensearch-project/OpenSearch/blob/main/SECURITY.md first, to avoid a false negative?

**Comment by xuezhou25** [comment]
The documentation for [ES 7.10.2](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/modules-http.html) might have some mistake(It is documented right on [ES 8.4](https://www.elastic.co/guide/en/elasticsearch/reference/8.4/modules-network.html)) about the option to include `error_trace` in the server response by setting the `http.detailed_errors.enabled: false`

When the config setting `http.detailed_errors.enabled: false`, the request of `error_trace=true` will return 400 error.
```
$ curl "http://localhost:9200/opensearch/_search?error_trace=false"
{"error":"IndexNotFoundException[no such index [opensearch]]","status":404}

$ curl "http://localhost:9200/opensearch/_search?error_trace=true" 
{"error":"error traces in responses are disabled.","status":400}   
```
This code regarding to this setting is from: https://github.com/opensearch-project/OpenSearch/blob/2.2.1/server/src/main/java/org/opensearch/rest/RestController.java#L368-L375

When there is no config setting for `http.detailed_errors.enabled` or `http.detailed_errors.enabled: true`, the response of URL parameter is:
```
$ curl "http://localhost:9200/opensearch/_search?error_trace=true&pretty"
{
  "error" : {
    "root_cause" : [
      {
        "type" : "index_not_found_exception",
        "reason" : "no such index [opensearch]",
        "index" : "opensearch",
        "resource.id" : "opensearch",
        "resource.type" : "index_or_alias",
        "index_uuid" : "_na_",
        "stack_trace" : "[opensearch] IndexNotFoundException[no such index [opensearch]]\n\tat org.opensearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.indexNotFoundException(IndexNameExpressionResolver.java:1067)\n\tat org.opensearch.cluster.metadata.IndexNameExpressionResolver$WildcardExpressionResolver.innerResolve(IndexNameExpressionResolver.java:1004)
```

I think this can solve your concern @rursprung What's your opinion?

**Comment by anasalkouz** [comment]
@xuezhou25 could you please add this to OpenSearch Documentation?

**Comment by xuezhou25** [comment]
Raised an issue on Documentation Repo to update the `error_trace` URL parameter description.

**Comment by rursprung** [comment]
thanks for checking this, @xuezhou25! i've verified that `http.detailed_errors.enabled: false` works as expected on OpenSearch 2.x.
i'm fine with closing this issue in favour of https://github.com/opensearch-project/documentation-website/issues/1235 - feel free to close it if you agree with this.

**Comment by rursprung** [comment]
small update: `http.detailed_errors.enabled` does more than just prevent the inclusion of the stack trace / the usage of the `error_trace` parameter: it also suppresses (sometimes vital) information from error messages. see https://github.com/opensearch-project/OpenSearch/issues/4545

**Comment by xuezhou25** [comment]
Closing this by tracking on [1235](https://github.com/opensearch-project/documentation-website/issues/1235)

**Comment by rursprung** [comment]
i've noticed in testing now that #4545 is hurting us quite a bit. could we maybe reopen this ticket here to look for an option which _only_ disables the stack trace but doesn't otherwise suppress error messages? or do you consider this a bug which should be resolved in #4545 so that also with the config option set the error messages are always returned?

**Comment by dblock** [comment]
@rursprung I say open a new one?
